src/ucp/core/ucp_rkey.c

```diff
@@ -400,23 +400,31 @@ void ucp_rkey_resolve_inner(ucp_rkey_h rkey, ucp_ep_h ep)
                                                     UCT_MD_MEM_TYPE_HOST,
                                                     config->key.amo_lanes, rkey,
                                                     0, &uct_rkey);
-    if (rkey->cache.amo_lane != UCP_NULL_LANE) {
+    amo_sw = (rkey->cache.amo_lane == UCP_NULL_LANE);
+    if (amo_sw) {
+        rkey->cache.amo_proto     = &ucp_amo_sw_proto;
+        rkey->cache.amo_rkey      = UCT_INVALID_RKEY;
+    } else {
+        rkey->cache.amo_proto     = &ucp_amo_basic_proto;
         rkey->cache.amo_rkey      = uct_rkey;
         rkey->cache.amo_proto     = &ucp_amo_basic_proto;
     }
 
-    /* If we use sw rma need to resolve destination endpoint in order to
+    /* If we use sw rma/amo need to resolve destination endpoint in order to
```

Comment:
at this point all lanes have to switch to SW amo

Reply:
if you mean the sw/hw atomicity issue, then it's enough to calculate the lane per rkey. two different atomic variables (e.g 1st in shmem heap, 2nd in .bss/.data), with diff rkeys, can use different lanes.